docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,372 @@
+# C++ code style guidelines
+
+We use C++14, but currently consider C++17 too immature for usage.
+
+This documents contains guidelines regarding:
+
+* File structure
+* Comments and documentation
+* Naming
+* Formatting
+* Typing
+* Namespaces
+* Structs
+* Classes
+* Error handling
+* Memory handling
+* Resource handling
+* Other
+
+
+## File structure
+
+* Header files: `.hpp`
+* Source files: `.cpp`
+* Use `#pragma once` in headers.
+* Split class definition across a header and source file by default.
+* Include header definitions in the same file.
+
+* Order of includes
+	1. Related header.
+	2. `C` library.
+	3. `C++` library.
+	4. Other libraries.
+	5. Fetch library.
```

Comment:
I would update to:

1. Internal library Includes
2. Other fetch library headers
3. Vendor library headers
4. System / standard library headers